File: ios/liquid_toasts/Sources/liquid_toasts/ToastImageDecoder.swift
import UIKit

/// Off-main image decoding for the leading avatar. Previously the bytes were
/// decoded synchronously inside the method-channel call (blocking the main
/// thread on large sources); now decode — and, for big images, downsampling —
/// happens on a background task and the pixels are attached to the model when
/// ready.
enum ToastImageDecoder {
  /// Fully decodes [data]; sources meaningfully larger than the avatar slot
  /// are thumbnailed to ~3× the 26 pt slot (Retina headroom) so a photo-sized
  /// payload never lives in memory at full resolution. Returns nil for
  /// undecodable bytes.
  static func decode(_ data: Data) async -> UIImage? {
    guard let image = UIImage(data: data) else { return nil }
    let pixelMax = max(image.size.width, image.size.height) * image.scale
    let targetPoints = ToastMetrics.avatarSize * 3
    if pixelMax > 256 {
      let scale = targetPoints / max(image.size.width, image.size.height)
      let size = CGSize(width: image.size.width * scale,
                        height: image.size.height * scale)
      return await image.byPreparingThumbnail(ofSize: size) ?? image
    }
    // Small source: just force the decode off the render path.
    return await image.byPreparingForDisplay() ?? image
  }
}

File: android/src/main/kotlin/com/rehmatsg/liquid_toasts/ToastImageDecoder.kt
package com.rehmatsg.liquid_toasts

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Off-main image decoding for the leading avatar — the Android analog of
 * `ToastImageDecoder.swift`. Sources meaningfully larger than the avatar slot
 * are downsampled to ~3× the 26dp slot (display headroom) so a photo-sized
 * payload never lives in memory at full resolution. Returns null for undecodable
 * bytes (the manager then collapses the reserved slot).
 */
internal object ToastImageDecoder {
    /** Only downsample when the source's larger dimension exceeds this (px). */
    private const val DOWNSAMPLE_THRESHOLD_PX = 256

    suspend fun decode(bytes: ByteArray, density: Float): Bitmap? = withContext(Dispatchers.Default) {
        // Bounds-only pass: read the source dimensions without allocating pixels.
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeByteArray(bytes, 0, bytes.size, bounds)
        val srcMax = maxOf(bounds.outWidth, bounds.outHeight)
        if (srcMax <= 0) return@withContext null // undecodable

        if (srcMax <= DOWNSAMPLE_THRESHOLD_PX) {
            // Small source: plain decode.
            return@withContext BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        }

        // Target ~3× the avatar slot in device pixels, mirroring the iOS Retina rule.
        val targetPx = (ToastMetrics.AVATAR_SIZE * 3 * density).toInt().coerceAtLeast(1)
        val opts = BitmapFactory.Options().apply {
            inSampleSize = computeSampleSize(srcMax, targetPx)
        }
        BitmapFactory.decodeByteArray(bytes, 0, bytes.size, opts)
    }

    /** Largest power-of-two sample size that keeps [srcMax] at or above [targetPx]. */
    private fun computeSampleSize(srcMax: Int, targetPx: Int): Int {
        var sample = 1
        while (srcMax / (sample * 2) >= targetPx) sample *= 2
        return sample
    }
}
